activesupport/lib/active_support/core_ext/module/delegation.rb

```diff
@@ -197,9 +197,34 @@ def delegate(*methods, to: nil, prefix: nil, allow_nil: nil, private: nil)
       method_name = prefix ? "#{method_prefix}#{method}" : method
       method_names << method_name.to_sym
 
+      begin
+        arity = \
+          if to.is_a?(Class)
+            to.method(method).arity
+          elsif receiver == "self.class"
+            method(method).arity
+          end
+      rescue NameError
+        # Do nothing. Just fall back to `...`
+      end
+
       # Attribute writer methods only accept one argument. Makes sure []=
       # methods still accept two arguments.
-      definition = /[^\]]=\z/.match?(method) ? "arg" : "..."
+      definition = \
+        if /[^\]]=\z/.match?(method)
+          "arg"
+        else
+          case arity
+          when 0
+            "&block"
+          when 1
+            "arg, &block"
```

Comment:
I don't think this is correct, because required keyword arguments have positive arity

```ruby
>> def foo(bar:) = nil
=> :foo
>> method(:foo).arity
=> 1
```

Would be good to add a test case for it.

Reply:
Hum, wait, there seem to be a test for it, I don't understand how it passes.

Reply:
Currently released Rubies still haven't achieved the complete separation of kwargs and Hash. It used to be in plan for 3.0, but the team gave it up because it turned out to be a huge incompatibility. I'm not sure if we're still aiming that direction, but if it's done, I mean, if kwargs is not treated as a Hash in Ruby, I bet current Rails codebase would completely stop working, and we have to add hundreds of curly brackets here and there.

Reply:
I don't follow what you mean:

```ruby
>> def foo(a:) = a
=> :foo
>> def bar(args) = foo(args)
=> :bar
>> bar(a: 1)
(irb):12:in `foo': wrong number of arguments (given 1, expected 0; required keyword: a) (ArgumentError)
```

Reply:
My bad. I thought I tested that pattern, but there was a mistake in the test code. I should have called instance methods in the test code here,

```diff
     assert_nothing_raised do
-      d.zero
-      d.one(1)
-      d.two(1, 2)
-      d.kwargs(a: 1, b: 2)
+      d.new.zero
+      d.new.one(1)
+      d.new.two(1, 2)
+      d.new.kwargs(a: 1, b: 2)
     end
```

and now I'm seeing:

```ArgumentError: wrong number of arguments (given 1, expected 0; required keywords: a, b)```

So, after all I updated the logic here to use `Method#parameters` rather than arity, in order to properly handle kwargs. The basic concept of the patch is unchanged, and although we could now investigate the details of kwargs, this time I made it much simpler as follows.
- If the method takes positional required arguments only (or zero arguments), the delegator is defined with the exact same argument names with the original methods' arguments
- If the method takes any kind of kwargs, the delegator is defined to take `...`
- The delegator is defined to take `&block` only when the original method takes `&block`

@byroot I'm sorry for the update, but could you take a look again?